File: jorney/zmq/parallel/csharp/worker/main/src/WorkerMain.cs
using System;
using System.Threading;
using ZeroMQ;

namespace Zmq.Parallel
{
  static partial class Worker
  {
    public static void Main(string[] args)
    {
			using (var context = new ZContext())
			using (var receiver = new ZSocket(context, ZSocketType.PULL))
			using (var sink = new ZSocket(context, ZSocketType.PUSH))
			{
				receiver.Connect("tcp://localhost:5557");
				sink.Connect("tcp://localhost:5558");
				while (true)
				{
					var replyBytes = new byte[4];
					receiver.ReceiveBytes(replyBytes, 0, replyBytes.Length);
					int workload = BitConverter.ToInt32(replyBytes, 0);
					Console.Write("{0} ", workload);
					Thread.Sleep(workload);
					sink.Send(new byte[0], 0, 0);
				}
			}
    }
  }
}

File: jorney/zmq/parallel/cpp/worker/main/src/WorkerMain.cpp
#include <string>
#include <iostream>
#include <sstream>
#include "zmq.hpp"
#include "basics/Thread.h"

int main() {
	zmq::context_t context(1);
	zmq::socket_t receiver(context, ZMQ_PULL);
	receiver.connect("tcp://localhost:5557");
	zmq::socket_t sender(context, ZMQ_PUSH);
	sender.connect("tcp://localhost:5558");
	while (true) {
		zmq::message_t message;
		int workload;
		receiver.recv(&message);
		std::string smessage(static_cast<char*>(message.data()), message.size());
		std::istringstream iss(smessage);
		iss >> workload;
		basics::msleep(workload);
		message.rebuild();
		sender.send(message);
		std::cout << "." << std::flush;
	}
	return 0;
}